WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt

```diff
@@ -98,33 +154,34 @@ class PrepublishingCategoriesViewModel @Inject constructor(
             return
         }
 
-        val categoryList = (uiState.value as? ContentUiState)?.selectedCategoryIds ?: return
+        val categoryList = getSelectedIds()
         updateCategoriesJob?.cancel()
+
         updateCategoriesJob = launch(bgDispatcher) {
             updatePostCategoriesUseCase.updateCategories(categoryList.toList(), editPostRepository)
         }
-        trackCategoriesChangedEvent()
     }
 
-    fun addSelectedCategory(categoryId: Long) {
-        uiState.value.let { state ->
-            if (state is ContentUiState) {
-                state.selectedCategoryIds.add(categoryId)
-                _uiState.value = state.copy()
-            }
-        }
+    private fun getSelectedIds(): List<Long> {
+        val uiState = uiState.value as UiState
+        return uiState.categoriesListItemUiState.filter { x -> x.checked }
+                .map { id -> id.categoryNode.categoryId }
+                .toList()
     }
 
-    fun removeSelectedCategory(categoryId: Long) {
-        uiState.value.let { state ->
-            if (state is ContentUiState) {
-                state.selectedCategoryIds.remove(categoryId)
-                _uiState.value = state.copy()
-            }
-        }
+    private fun getSiteCategories() =
+            getCategoriesUseCase.getSiteCategories(siteModel)
+
+    private fun getPostCategories() =
+            getCategoriesUseCase.getPostCategories(editPostRepository, siteModel)
+
+    private fun hasChanges(): Boolean {
+        val postCategories = getPostCategories()
+        val stateSelectedCategories = getSiteCategories()
+        return (stateSelectedCategories != postCategories)
```

Comment:
It seems we are comparing two different data types + the comparison is order dependant, is it intentional?

Reply:
Nope, that was "you've been at the computer for 10 hours on a Sunday" ... and that's why we have code reviews. Thanks! Fixed in a983929

Reply:
:D